### PR TITLE
Recurring orders testing

### PR DESCRIPTION
**Pull request by kuba-end**

### CODE REVIEW

File: spec/Form/Transformer/MollieIntervalTransformerSpec.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        $this->shouldHaveType(MollieIntervalTransformer::class);
+    }
+
+    function it_should_implement_data_transformer_interface(): void
+    {
+        $this->shouldImplement(DataTransformerInterface::class);
+    }
+
+    function it_transform_days(): void
```
Comment (senghe) [inline spec/Form/Transformer/MollieIntervalTransformerSpec.php:27]:
```suggestion
    function it_transforms_days(): void
```

File: spec/Form/Transformer/MollieIntervalTransformerSpec.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        $this->transform($value)->shouldReturn([
+            0 => "12 months",
+            'amount' => "12",
+            1 => "12",
+            'step' => "months",
+            2 => "months",
+        ]);
+    }
+
+    function it_return_empty_array_when_unsupported_format(): void
```
Comment (senghe) [inline spec/Form/Transformer/MollieIntervalTransformerSpec.php:63]:
```suggestion
    function it_returns_empty_array_when_meets_unsupported_format(): void
```

File: spec/Generator/SubscriptionScheduleGeneratorSpec.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        $this->shouldHaveType(SubscriptionScheduleGenerator::class);
+    }
+
+    function it_should_implement_subscription_shecdule_generator_interface(): void
+    {
+        $this->shouldImplement(SubscriptionScheduleGeneratorInterface::class);
+    }
+
+    function it_generates(
```
Comment (senghe) [inline spec/Generator/SubscriptionScheduleGeneratorSpec.php:43]: generates what?

File: spec/Grid/Filter/MollieSubscriptionStateSpec.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    ): void {
+        $data = ['not state' => []];
+        $source->getExpressionBuilder()->willReturn($builder);
+
+        $this->apply($source,'name', $data, []);
+
+        $source->restrict($builder->getWrappedObject()->in('not state', $data['not state']))->shouldNotBeCalled();
+    }
+
+    function it_returnes_when_array_key_has_empty_value(
```
Comment (senghe) [inline spec/Grid/Filter/MollieSubscriptionStateSpec.php:53]:
```suggestion
    function it_returns_when_array_key_has_empty_value(
```

File: spec/Menu/MollieMenuListenerSpec.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+use Sylius\Bundle\UiBundle\Menu\Event\MenuBuilderEvent;
+
+final class MollieMenuListenerSpec extends ObjectBehavior
+{
+    function it_is_initializable(): void
+    {
+        $this->shouldHaveType(MollieMenuListener::class);
+    }
+
+    function it_builds_menu(
```
Comment (senghe) [inline spec/Menu/MollieMenuListenerSpec.php:23]: It's not something we can test on this level

File: spec/Menu/MollieRecurringMenuListenerSpec.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+use Sylius\Bundle\UiBundle\Menu\Event\MenuBuilderEvent;
+
+final class MollieRecurringMenuListenerSpec extends ObjectBehavior
+{
+    function it_is_initializable(): void
+    {
+        $this->shouldHaveType(MollieRecurringMenuListener::class);
+    }
+
+    function it_builds_menu(
```
Comment (senghe) [inline spec/Menu/MollieRecurringMenuListenerSpec.php:23]: It's not something we want to test on this level

File: spec/Order/OrderPaymentRefundSpec.php [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+        $order->getPayments()->willReturn(new ArrayCollection([
+            $payment->getWrappedObject()
+        ]));
+        $payment->getMethod()->willReturn($paymentMethod);
+
+        $paymentMethod->getGatewayConfig()->willReturn($config);
+        $config->getFactoryName()->willReturn('mollie_subscription_test');
+    }
+    function it_is_initializable(): void
```
Comment (senghe) [inline spec/Order/OrderPaymentRefundSpec.php:67]: No new line

File: spec/Order/ShipmentUnitClonerSpec.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        $this->shouldHaveType(ShipmentUnitCloner::class);
+    }
+
+    function it_should_have_implement_shipment_unit_cloner_interface(): void
+    {
+        $this->shouldImplement(ShipmentUnitClonerInterface::class);
+    }
+
+    function it_clones(
```
Comment (senghe) [inline spec/Order/ShipmentUnitClonerSpec.php:34]: Clones what? :)

File: spec/Order/SubscriptionOrderClonerSpec.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    {
+        $this->shouldHaveType(SubscriptionOrderCloner::class);
+    }
+
+    function it_should_implements_subscription_order_cloner_interface(): void
+    {
+        $this->shouldImplement(SubscriptionOrderClonerInterface::class);
+    }
+
+    function it_clones(
```
Comment (senghe) [inline spec/Order/SubscriptionOrderClonerSpec.php:63]: Clones what? 🙃

File: spec/Payments/MethodResolver/MolliePaymentMethodResolverSpec.php [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        $order->hasRecurringContents()->willReturn(true);
+        $paymentMethod->getGatewayConfig()->willReturn($gatewayConfig);
+        $gatewayConfig->getFactoryName()->willReturn('not_mollie_subscription');
+
+        $mollieMethodFilter->recurringFilter([$parentMethod])->willReturn([$parentMethod]);
+
+        $this->getSupportedMethods($subject)->shouldReturn([$parentMethod]);
+    }
+
+    function it_supports(
```
Comment (senghe) [inline spec/Payments/MethodResolver/MolliePaymentMethodResolverSpec.php:147]: Supports what? When?

File: spec/Provider/Form/ResolverGroupProviderSpec.php [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        $form->getData()->willReturn($data);
+        $data->isRecurring()->willReturn(true);
+
+        $this->provide($form)->shouldReturn([
+            'sylius',
+            'recurring_product_variant'
+        ]);
+    }
+
+    function it_provides_when_non_recurring(
```
Comment (senghe) [inline spec/Provider/Form/ResolverGroupProviderSpec.php:41]: Provides what?

File: src/Action/Api/CreateMollieSubscriptionAction.php [hunk 12]
```diff
@@ -0,0 +1,110 @@
+<?php
+declare(strict_types=1);
```
Comment (senghe) [inline src/Action/Api/CreateMollieSubscriptionAction.php:2]: No OS header

File: src/Action/Api/CreateMollieSubscriptionAction.php [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+use Payum\Core\ApiAwareInterface;
+use Payum\Core\Bridge\Spl\ArrayObject;
+use Payum\Core\Exception\RequestNotSupportedException;
+use Payum\Core\GatewayAwareInterface;
+use Payum\Core\GatewayAwareTrait;
+use Payum\Core\Request\GetCurrency;
+use Payum\Core\Security\GenericTokenFactoryAwareInterface;
+use Payum\Core\Security\GenericTokenFactoryInterface;
+
+final class CreateMollieSubscriptionAction extends BaseApiAwareAction implements ActionInterface, ApiAwareInterface, GatewayAwareInterface, GenericTokenFactoryAwareInterface
```
Comment (senghe) [inline src/Action/Api/CreateMollieSubscriptionAction.php:22]: Why it's added here? It's not a spec. Could you review it?

File: src/Action/CaptureAction.php [hunk 14]
```diff
@@ -65,40 +64,34 @@ public function execute($request): void
...
-                );
-
-                $details['cancel_token'] = $cancelToken->getHash();
-                $this->gateway->execute(new CreateCustomer($details));
-                $this->gateway->execute(new CreateInternalRecurring($details));
-                $this->gateway->execute(new CreateOnDemandSubscription($details));
-            } elseif ('recurring' === $details['sequenceType']) {
-                $this->gateway->execute(new CreateOnDemandSubscriptionPayment($details));
-            }
+            $cancelToken = $this->tokenFactory->createToken(
```
Comment (senghe) [inline src/Action/CaptureAction.php:68]: Why this change?

File: src/Creator/MollieMethodsCreator.php [hunk 15]
```diff
@@ -42,24 +37,20 @@ final class MollieMethodsCreator implements MollieMethodsCreatorInterface
...
+        MollieMethodsCreatorHelperInterface $helper
     )
     {
-        $this->methodsFactory = $methodsFactory;
-        $this->entityManager = $entityManager;
         $this->loggerAction = $loggerAction;
         $this->mollieApiClient = $mollieApiClient;
         $this->gatewayConfigRepository = $gatewayConfigRepository;
-        $this->factory = $factory;
+        $this->helper = $helper;
```
Comment (senghe) [inline src/Creator/MollieMethodsCreator.php:53]: We have to avoid names like "Helper"

File: src/Form/Type/MollieGatewayConfigType.php [hunk 16]
```diff
@@ -32,6 +32,7 @@
 use Symfony\Component\Form\FormBuilderInterface;
 use Symfony\Component\Form\FormEvent;
 use Symfony\Component\Form\FormEvents;
+use function Clue\StreamFilter\fun;
```
Comment (senghe) [inline src/Form/Type/MollieGatewayConfigType.php:35]: What's this?

File: src/Form/Type/MollieIntervalType.php [hunk 17]
```diff
@@ -24,31 +22,15 @@ public function __construct(DataTransformerInterface $transformer)
...
-                new NotBlank([
-                    'groups' => ['recurring_product_variant'],
-                ]),
-                new GreaterThan([
-                    'value' => 0,
-                    'groups' => ['recurring_product_variant'],
-                ])
-            ]
-        ]);
+        $builder->add('amount', NumberType::class);
```
Comment (senghe) [inline src/Form/Type/MollieIntervalType.php:25]: Why changes in this files?

File: src/Form/Type/MollieSubscriptionGatewayConfigurationType.php [hunk 18]
```diff
@@ -12,9 +12,21 @@
...
+use Symfony\Component\Form\Extension\Core\Type\NumberType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Validator\Constraints\NotBlank;
+use Symfony\Component\Validator\Constraints\Range;
+use Symfony\Component\Validator\Constraints\Regex;
 
 final class MollieSubscriptionGatewayConfigurationType extends AbstractType
 {
+    public function buildForm(FormBuilderInterface $builder, array $options): void
```
Comment (senghe) [inline src/Form/Type/MollieSubscriptionGatewayConfigurationType.php:24]: What's this?

File: src/Migrations/Version20220112141532.php [hunk 19]
```diff
@@ -20,7 +20,7 @@ public function getDescription(): string
     public function up(Schema $schema): void
     {
         // this up() migration is auto-generated, please modify it to your needs
-        $this->addSql('CREATE TABLE bitbag_mollie_subscription_orders (subscription_id INT NOT NULL, order_id INT NOT NULL, INDEX IDX_DCE71BD39A1887DC (subscription_id), INDEX IDX_DCE71BD38D9F6D38 (order_id), PRIMARY KEY(subscription_id, order_id)) DEFAULT CHARACTER SET UTF8 COLLATE `UTF8_unicode_ci` ENGINE = InnoDB');
+//        $this->addSql('CREATE TABLE bitbag_mollie_subscription_orders (subscription_id INT NOT NULL, order_id INT NOT NULL, INDEX IDX_DCE71BD39A1887DC (subscription_id), INDEX IDX_DCE71BD38D9F6D38 (order_id), PRIMARY KEY(subscription_id, order_id)) DEFAULT CHARACTER SET UTF8 COLLATE `UTF8_unicode_ci` ENGINE = InnoDB');
```
Comment (senghe) [inline src/Migrations/Version20220112141532.php:23]: Why?

**Review by senghe — CHANGES_REQUESTED** [review]
Can you check if behats are failing or not in this PR?

File: src/Action/Api/CreateMollieSubscriptionAction.php (hunk 12 above)
Comment (kuba-end) [reply src/Action/Api/CreateMollieSubscriptionAction.php:2]: //there is no such file in recurring branch
